### DISCUSSION TITLE
Parallelizing download

### DISCUSSION THREAD

**Discussion by agenteo** [icloud-photos-downloader/icloud_photos_downloader#1023] [Ideas]
Hi, have you already explored parallelizing the files download? What did you learn? Cheers

**Comment by AndreyNikiforov** [comment]
> Hi, have you already explored parallelizing the files download? What did you learn? Cheers

What bottleneck will it fix? What evidence is there to confirm it is indeed a bottleneck?

**Reply by agenteo** [reply]
I've been using the app for a couple of days and it's great but with my huge collection of tens of thousands of photos/videos it will take a couple of weeks to download.

I didn't spend much time looking for evidence but I don't think the time it's taking is about my network (it might be capped on Apple' side), or my CPU or memory. I just assumed fetching those images one at the time is slower than if the script batched and fetched them in parallel.

**Reply by AndreyNikiforov** [reply]
> I've been using the app for a couple of days and it's great but with my huge collection of tens of thousands of photos/videos it will take a couple of weeks to download.
> 
> I didn't spend much time looking for evidence but I don't think the time it's taking is about my network (it might be capped on Apple' side), or my CPU or memory. I just assumed fetching those images one at the time is slower than if the script batched and fetched them in parallel.

You are on the right path. If you figure out download speed and compare it to your ISP speed tests, I bet you will see that you are maxing out your download throughput. If it is true, doing parallel downloading will not help, but may even hurt (depending what you care about).

**Reply by agenteo** [reply]
Yeah I know what you mean. And that's not the case. My NIC Rx is around 300KB/s. I have a laptop dedicated to this download. And yes my ISP is above that. I bet Apple caps the download.

I had the same slow speed when I was manually downloading from iCloud years ago. One of the reasons I moving my stuff off iCloud.

Going back to my initial question about parallelizing. If someone knows the codebase it could be a fun experiment to test. I am a bit opaque on how this script works, ie. using an Apple API? Scraping iCloud?

**Reply by AndreyNikiforov** [reply]
> Going back to my initial question about parallelizing. If someone knows the codebase it could be a fun experiment to test.

See #191 and search for `thread` in PRs & issues.

**Reply by AndreyNikiforov** [reply]
As a suggestion for your testing - run two instances of `icloudpd`. If combined speed ~doubles, it will support parallelizing downloading.

Also note, that downloading huge collection once is not the primary use case. Yes, everybody starts there, but after initial download it is only incremental updates that matter. If your `icloudpd` runs as a service and checks for updates every, say, 10min, then increased download speed would not have much impact.

**Reply by agenteo** [reply]
> > Going back to my initial question about parallelizing. If someone knows the codebase it could be a fun experiment to test.
> 
> See #191 and search for `thread` in PRs & issues.
Thanks for pointing to this. That's exactly the level of detail I was looking for. I must have missed when I searched the archives :)

And thanks for the tip about starting a new instance.


> Also note, that downloading huge collection once is not the primary use case. Yes, everybody starts there, but after initial download it is only incremental updates that matter. If your `icloudpd` runs as a service and checks for updates every, say, 10min, then increased download speed would not have much impact.

Yes that makes sense. And Apple clearly doesn't make it easy for people to get their data out.

**Reply by agenteo** [reply]
Just leaving a note for posterity. I tried to run 2 icloudpd on my account but I ended up getting an IO error that mentioned running out of space (I wasn't). It looks like both instances were downloading the same file. It was worth a shot.
